Added method to set global MNI template resolution. Added 3mm templat…

ejolly opened the pull request:
…es and ROI parcellation to resources.

ljchang commented:
Thanks for this.  Here are a couple of things.

1) Tests are failing.  I think because the mask size has changed the number of voxels.  We need to fix this.

2) Make sure you add tests for the new pref module.

3) I don't think it's a great idea to add the parcellations to the repository for several reasons.  Mostly because we should minimize including binary files and the paper isn't published yet.  If they are integral to something that would be different.  Instead I would recommend adding them to a neurovault repository and just loading them from the web if you need them for a specific thing.  I'll add you as a collaborator to my collection I'm using for this.

ljchang commented:
@ejolly: I added you as contributor to neurosynth parcellation collection.  you should be able to write to that collection now.

I think tests should probably be run at both resolutions.  This will ensure that we don't miss a method that will go back to 2mm default.  Okay if it takes longer.  better to find out in test than later.

ejolly commented:
So looks like the test for region masking and thresholding is failed but *only on python 3*. I think it has something to do with how `create_sphere()` is making the test data, which is causing the evaluation of what should be a binary mask (2 unique) values, to seem like it has 3 unique values. 

@jcheong0428 Looks like you worked a bunch of the `create_sphere()` functionality. Can you take a look at how you hard-coded stuff and make sure that it's not dependent on assuming a 2mm space/mask, and is modular enough to handle 3mm? And also that the way things are implemented can handle both python 2 and 3?

ljchang commented:
create_sphere is a mess.  I think we should just rewrite it.  It's not working in Python3 and others have been having issues with it see issue #135.  Looking into stealing nilearn's version.

ljchang commented:
@ejolly: want to resolve conflicts and see if Python3 works now?  It should be just about ready to merge

ejolly commented:
So I made a bunch of changes with the most recent version of nltools (0.3.1) release that also fix the issue @kristinarapuano was having. This should resolve most everything, but for the life of me I can't figure out how to get `test_data` to pass on python 3. It fails on line:
```
 r = mask.regions(min_region_size=10)
 assert len(np.unique(r.to_nifti().get_data())) == 2 #this resolves to 3 instead of 2 on travis
```

First i thought it was 3mm vs 2mm template issue, but the test passes with 3mm template on python 2! It also passes for me locally using a python 3 environment getting the right value. So I have no idea why travis is getting a totally different answer. There's 2 possibilities I can think of:

1. it's the way the `Simulator` is generating data. Maybe theres an edge simulator case to handle here? 
2. The brain masks have been updated in my PR relative to the master, including the 2mm. After a previous discussion we had, I'm including MNI templates that 0-out ventricles, whereas the ones that have been included thus far don't, leaving fewer voxels and differing shapes for a lot of tests. Mine seem to work fine since I updated the tests, but is it possible travis is using a cached version of the package resources?

@ljchang Any suggestions?

ljchang commented:
@ejolly:  One thing that is making it hard for to debug the problem is that create_mask no longer plots.  This is an issue with the new version of create_mask, not your code.  Any chance you have an idea for how to fix this?

`python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.pyc in __call__(self, obj)
    332                 pass
    333             else:
--> 334                 return printer(obj)
    335             # Finally look for special method names
    336             method = get_real_method(obj, self.print_method)

/usr/local/lib/python2.7/site-packages/IPython/core/pylabtools.pyc in <lambda>(fig)
    239 
    240     if 'png' in formats:
--> 241         png_formatter.for_type(Figure, lambda fig: print_figure(fig, 'png', **kwargs))
    242     if 'retina' in formats or 'png2x' in formats:
    243         png_formatter.for_type(Figure, lambda fig: retina_figure(fig, **kwargs))

/usr/local/lib/python2.7/site-packages/IPython/core/pylabtools.pyc in print_figure(fig, fmt, bbox_inches, **kwargs)
    123 
    124     bytes_io = BytesIO()
--> 125     fig.canvas.print_figure(bytes_io, **kw)
    126     data = bytes_io.getvalue()
    127     if fmt == 'svg':

/usr/local/lib/python2.7/site-packages/matplotlib/backend_bases.pyc in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, **kwargs)
   2206                     orientation=orientation,
   2207                     dryrun=True,
-> 2208                     **kwargs)
   2209                 renderer = self.figure._cachedRenderer
   2210                 bbox_inches = self.figure.get_tightbbox(renderer)

/usr/local/lib/python2.7/site-packages/matplotlib/backends/backend_agg.pyc in print_png(self, filename_or_obj, *args, **kwargs)
    505 
    506     def print_png(self, filename_or_obj, *args, **kwargs):
--> 507         FigureCanvasAgg.draw(self)
    508         renderer = self.get_renderer()
    509         original_dpi = renderer.dpi

/usr/local/lib/python2.7/site-packages/matplotlib/backends/backend_agg.pyc in draw(self)
    428             if toolbar:
    429                 toolbar.set_cursor(cursors.WAIT)
--> 430             self.figure.draw(self.renderer)
    431         finally:
    432             if toolbar:

/usr/local/lib/python2.7/site-packages/matplotlib/artist.pyc in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python2.7/site-packages/matplotlib/figure.pyc in draw(self, renderer)
   1293 
   1294             mimage._draw_list_compositing_images(
-> 1295                 renderer, self, artists, self.suppressComposite)
   1296 
   1297             renderer.close_group('figure')

/usr/local/lib/python2.7/site-packages/matplotlib/image.pyc in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    136     if not_composite or not has_images:
    137         for a in artists:
--> 138             a.draw(renderer)
    139     else:
    140         # Composite any adjacent images together

/usr/local/lib/python2.7/site-packages/matplotlib/artist.pyc in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python2.7/site-packages/matplotlib/axes/_base.pyc in draw(self, renderer, inframe)
   2397             renderer.stop_rasterizing()
   2398 
-> 2399         mimage._draw_list_compositing_images(renderer, self, artists)
   2400 
   2401         renderer.close_group('axes')

/usr/local/lib/python2.7/site-packages/matplotlib/image.pyc in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    136     if not_composite or not has_images:
    137         for a in artists:
--> 138             a.draw(renderer)
    139     else:
    140         # Composite any adjacent images together

/usr/local/lib/python2.7/site-packages/matplotlib/artist.pyc in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python2.7/site-packages/matplotlib/image.pyc in draw(self, renderer, *args, **kwargs)
    546         else:
    547             im, l, b, trans = self.make_image(
--> 548                 renderer, renderer.get_image_magnification())
    549             if im is not None:
    550                 renderer.draw_image(gc, l, b, im)

/usr/local/lib/python2.7/site-packages/matplotlib/image.pyc in make_image(self, renderer, magnification, unsampled)
    772         return self._make_image(
    773             self._A, bbox, transformed_bbox, self.axes.bbox, magnification,
--> 774             unsampled=unsampled)
    775 
    776     def _check_unsampled_image(self, renderer):

/usr/local/lib/python2.7/site-packages/matplotlib/image.pyc in _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification, unsampled, round_to_pixel_border)
    368                 # old versions of numpy do not work with `np.nammin`
    369                 # and `np.nanmax` as inputs
--> 370                 a_min = np.ma.min(A).astype(scaled_dtype)
    371                 a_max = np.ma.max(A).astype(scaled_dtype)
    372                 # scale the input data to [.1, .9].  The Agg

/usr/local/lib/python2.7/site-packages/numpy/ma/core.pyc in astype(self, newtype)
   3203                 output._mask = self._mask.astype([(n, bool) for n in names])
   3204         # Don't check _fill_value if it's None, that'll speed things up
-> 3205         if self._fill_value is not None:
   3206             output._fill_value = _check_fill_value(self._fill_value, newtype)
   3207         return output

AttributeError: 'MaskedConstant' object has no attribute '_fill_value'

<matplotlib.figure.Figure at 0x10312b650>
`

ejolly commented:
@ljchang This looks like a bug with the latest version of matplotlib. They're supposed to update their release to handle this issue soonish. Less than ideal workaround is to enforce usage of matplotlib==2.0.0, which doesn't have this issue.

Here's the page:
https://github.com/matplotlib/matplotlib/issues/9280

ljchang commented:
nice find @ejolly.  I guess we can wait it out.  only seems to be affecting the create mask plots at the moment.

ejolly commented:
Actually found it affects another use case: copy a brain data instance and replacing its internal data values with something else. If those internal values are mostly zeros, it tends to fail e.g.

```
weight_map = dat.copy()
weight_map.data = np.zeros_like(dat.data)
weight_map.data[certain_voxels] = my_values # len(certain_voxels) or my_values ~ 5000

weight_map.plot() #fails
```

ejolly commented:
Just added 3 different kinds of robust standard error estimators to `.regress()` (i.e. sandwich estimators) including one that's commonly used for robustness to serial auto-correlation in the forecasting literature when an explicit ARMA model isn't calculated. Validated against statsmodels and stata implementations. Still need some unit tests on them though. 

Yea an output class could be useful for organizing things a bit more.

ljchang commented on nltools/data.py:
```diff
@@ -1410,13 +1410,13 @@ def aggregate(self, mask, func):
-    def threshold(self, thresh=0, binarize=False):
-        '''Threshold Brain_Data instance
+    def threshold(self, thresh=0, binarize=False, upper_only=False):
+        '''Threshold Brain_Data instance. If data contain both positive and negative values, defaults to two-sided thresholding automatically. This can be changed with the upper_only flag.
```
This is tricky as this function is used in a bunch of different ways.  it's currently being used for thresholding masks at a certain percentile and is intended to be one sided.  I don't think we should change the defaults as we will have to fix every function that calls it.  I can see a case to be made for having a threshold function for statistics, but we already have one (nltools.stats.threshold).

Maybe add something like a 'side=upper' flag with options for ['upper','lower','both'].

I don't think it is a good idea to have a function default change based on the content of the data in this context.  It will likely lead to unintended behavior.

ejolly replied:
Fixed as per what we discussed the other day, along with test added

ejolly commented:
@ljchang Think we should be good to go all checks seem to pass on Travis and locally for me; seems like it was mostly an issue with ventricle vs no-ventricle masks. Masks with ventricles are default now (as before), but added a preference option to switch them out easily if the user wants using the new global MNI settings.  Added an additional test for two-sided thresholding too and fixed a bug in your bootstrap function.

ljchang commented on nltools/data/brain_data.py:
```diff
@@ -83,11 +86,17 @@
 from scipy.spatial.distance import squareform
 import six
-from sklearn.metrics.pairwise import pairwise_distances
+from sklearn.metrics.pairwise import (pairwise_distances,
```
Not sure what happened here, but we only need pairwise_distances from sklearn.  

transform_pairwise, and summarize_bootstrap should be in nltools.stats module

ejolly replied:
yea I fixed this already. GH is just showing history cause it's broken up into a bunch of commits on this PR.

ljchang commented on nltools/stats.py:
```diff
@@ -26,7 +26,8 @@
             '_hc0',
             '_hc3',
-            '_hac']
+            '_hac',
```
can you name these something more descriptive?

Maybe something like _robust_estimator_hac?

ejolly replied:
Done

ljchang commented:
@ejolly.  Went through most of your code, but couldn't find the bug you found with bootstrap.  What was it?

ljchang commented on nltools/tests/test_data.py:
```diff
@@ -67,7 +67,7 @@ def test_brain_data_3mm(tmpdir):
     assert len(dat[index]) == len(index)
     index = dat.Y == 1
-    assert len(dat[index]) == index.values.sum()
+    assert len(dat[index.values.flatten()]) == index.values.sum()
```
Let's keep the original test too as I want it to work with both pandas and numpy logical indexing.

add this back in `assert len(dat[index]) == index.values.sum()`

ejolly replied:
Added back in. Also made sure all tests in 3mm are also in 2mm and visa versa. Looks like some were missing across the different versions.

ljchang commented:
We also need to update a tutorial on how to use preference class, but let's save it for hackathon day.

ejolly commented:
@ljchang Bootstrap bug was just a variable naming issue in this [line](https://github.com/ljchang/nltools/blob/master/nltools/stats.py#L559)
You have a variable called 'a', but you meant 'wz' as you created that from a copy in the line above and use that in line 561, but dont actually add data to it, but instead add data to 'a'.

ejolly commented:
@ljchang Ok think we should _actually_ be good to go this time.

ljchang commented:
haha. ok.  let me go through the new stuff.  thanks for catching that bug.  surprised I didn't see it.  I don't think there is an explicit unit test for summarize bootstrap yet, but the function is implicitly tested when bootstrap is called as it is always used in the output.
